### PR TITLE
devices: Config for the Fairphone 4

### PR DESCRIPTION
**Pull request by z3ntu**
Supercedes #193

### CODE REVIEW

**Comment by maciek134** [comment]
After flashing it fails to reboot into `fastbootd` - blank but turned on screen. UT recovery boots fine, but also can't reboot into `fastbootd` from there. Any idea what might be wrong? Ofc using the latest installer.

**Comment by z3ntu** [comment]
@maciek134 Correct, working on fixing this right now :)

**Comment by z3ntu** [comment]
Recovery is now built with https://github.com/Halium/android_device_halium_halium_arm64/pull/18

It installs fine now (incl. system image stuff in recovery) but doesn't boot up afterwards. Still need to find out what is happening there and debug.

**Comment by maciek134** [comment]
Had the same yesterday, try not removing vendor_b and system_b, it booted
for me afterwards.

**Comment by z3ntu** [comment]
Seems `odm_a` partition is actually required, now it works.

Tested now by flashing FP4-FP3W-factory.zip, then flashing with all checkboxes ticked. Device boots up as expected. Then another flashing round over it with no checkboxes ticked and device still boots up as expected.

(and thanks to whoever for approving CI run!)

Three FIXME remaining, will open discussions on them, feedback welcome!

File: v2/devices/FP4.yml [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+                    sum: "c1ab7a12e48edddc089cd93265e475ff58242919427520a589fc19203c8fb628"
+                    algorithm: "sha256"
+                - url: "https://files.beidl.guru/fp4/dtbo.img"
+                  checksum:
+                    sum: "68ad83e20eb2b47592c2b7fe48161893770b1727c7b7ce7240504eaf8612d733"
+                    algorithm: "sha256"
+                - url: "https://public.z3ntu.xyz/tmp/recovery-fastbootd.img"
+                  checksum:
+                    sum: "ae7c97ca17fa93cfefd20c25e09d80f03b4921fe30f62f157ef85702a892ed53"
+                    algorithm: "sha256"
```
Comment (z3ntu) [inline v2/devices/FP4.yml:98]: These images should be hosted on cdimage, and maybe built on a CI system if possible?

Anyone know if they can be compressed and unpacked by the installer? Both boot and recovery partitions are 96MB on this device and therefore contain a lot of free space

File: v2/devices/FP4.yml [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  fastbootd:
+    title: "Reboot to fastbootd mode"
+    description: 'With the device still at the "FastBoot Mode", use the Volume buttons to switch to "Recovery Mode" and push the power button to confirm your selection. Then select "Fastboot mode" from the list.'
+    image: "phone_power_up"
+    button: true
+unlock:
+  - "confirm_model"
+  - "confirm_os"
+  - "unlock_phone"
+# FIXME bootloader_locked? https://www.fairphone.com/en/bootloader-unlocking-code-for-fairphone-3
```
Comment (z3ntu) [inline v2/devices/FP4.yml:42]: The `fastboot:oem_unlock` function allows providing a URL but 1. the code entry page with that is completely misleading on FP4; 2. FP4 would use `fastboot:flashing_unlock` and that doesn't use an URL

I already put "Unlock the bootloader" into the manual confirmation before flashing and expect users to follow that, I guess that's enough?

**Comment by rubencarneiro** [comment]
Can this be merged?

**Comment by amartinz** [comment]
> Can this be merged?

not yet, see Luca's latest comments.

Needs to be moved to cdimage servers at least.
I'll open a ticket for that to happen.

File: v2/devices/FP4.yml (hunk 2 above)
Comment (maciek134) [reply v2/devices/FP4.yml:42]: `fastboot:assert_var` could be used to make it more user friendly and actually check if the bootloader is unlocked? Or at least the error will be nicer (`Assertion error: expected bootloader variable unlocked to be yes but got no`)

Comment (z3ntu) [reply v2/devices/FP4.yml:42]: Sounds like a good idea, will try to implement this next week

**Comment by ajsalminen** [comment]
Tested but for some reason I'm not able to get a system booting after installation with this. Starting from a freshly flashed factory image and running UBports installer 0.9.6-beta on Windows with every checkbox checked.

Installer finishes, phone comes up with a message "Installing system update" and after that finishes it won't boot just showing the Fairphone powered by android screen).

No difference if I take out the parts deleting vendor_b and system_b from config and redo as @maciek134 suggested.

File: v2/devices/FP4.yml (hunk 2 above)
Comment (z3ntu) [reply v2/devices/FP4.yml:42]: @maciek134 I guess it looks well enough like this?

image


Doesn't show the `link` though that's configured in `unlock_phone`?

```yml
      - actions:
          - fastboot:assert_var:
              variable: "unlocked"
              value: "yes"
        fallback:
          - core:user_action:
              action: "unlock_phone"
```

**Comment by Flohack74** [comment]
@z3ntu whats missing here? More testing only? Our sponsor wants a report when this can be published :P

**Comment by z3ntu** [comment]
@Flohack74 The open discussions should be resolved, so pushing the current images to cdimage (I think you can do this?) and feedback about the unlock thingy. Then it's ready from my side

**Comment by Flohack74** [comment]
please send me all binaries you need to be uploaded on telegram :)

**Comment by ajsalminen** [comment]
As a followup to my testing:

I was now able to get a bootable install with this for unclear reasons. It could have been because I used TWRP to wipe userdata and skipped doing that with the installer. Wasn't an isolated test so I did some other things in the interim too like flash a different boot.img before running the installer and the installer downloaded the files again so maybe something changed with those...

All in all I suspect it could have had something to do with that formatting with TWRP. I think there's a chance that there's either an edge case or even something that affects people with a pristine device that only had the stock install on it previously like I did.

File: v2/devices/FP4.yml [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                  checksum:
+                    sum: "c1ab7a12e48edddc089cd93265e475ff58242919427520a589fc19203c8fb628"
+                    algorithm: "sha256"
+                - url: "https://cdimage.ubports.com/devices/FP4/dtbo.img"
+                  checksum:
+                    sum: "68ad83e20eb2b47592c2b7fe48161893770b1727c7b7ce7240504eaf8612d733"
+                    algorithm: "sha256"
+                - url: "https://cdimage.ubports.com/devices/FP4/recovery.img"
+                  checksum:
+                    sum: "ae7c97ca17fa93cfefd20c25e09d80f03b4921fe30f62f157ef85702a892ed53"
```
Comment (Flohack74) [inline v2/devices/FP4.yml:96]: Needs new checksum: 25b9baabde1aaa1b942fb43c3dc8c07954d8d11151b2c81c71c6e5ea33c3c00f

File: v2/devices/FP4.yml (hunk 2 above)
Comment (z3ntu) [reply v2/devices/FP4.yml:42]: Fixed with https://github.com/ubports/ubports-installer/pull/2798 but I don't consider this to be a blocker for the config.

File: v2/devices/FP4.yml (hunk 3 above)
Comment (amartinz) [reply v2/devices/FP4.yml:96]: Needs a new image with https://github.com/Halium/hybris-patches/pull/59 applied.

**Comment by mavaa** [comment]
Hello, I had the exact same issue as @ajsalminen on a fresh out of the box FP4 (yes, I'm very eager to test 😅 )

I used the newest beta of ubports-installer (0.9.6_beta-2), and the FP4.yml from the latest commit of this PR. Got the warning about incorrect checksum, which I ignored like a madman, and experienced the exact same thing as @ajsalminen described above. I haven't tried twrp or upgrading the bootloader yet, but I could try some of it in any order if you would like. I'm in no rush to get this working (except I really dislike my current daily driver phone 🙂 )

**Comment by Flohack74** [comment]
Guys please stay patient we are fixing this up now, and it will get merged if it works ;)

**Comment by Flohack74** [comment]
Ok latest config should work pretty well, who wants to test it?

**Comment by mavaa** [comment]
I did, and it worked completely without issue! Tested booting, some fiddling around and sending/receiving sms so far.

**Comment by z3ntu** [comment]
My test also worked fine now, on top of stock install with FP4-FP3W-factory.zip package.

Electron UI crapped itself as always but install still worked ;)

**Review by amartinz — APPROVED** [review]

**Comment by maciek134** [comment]
Did anyone test stock + latest updates? Just had an idea that we all flash
a pretty old version before trying ;p (I can try in a sec)

**Comment by z3ntu** [comment]
It seems there's now FP4-FP40-factory.zip available at https://support.fairphone.com/hc/en-us/articles/4405858261777-FP4-Install-Fairphone-OS-offline which is the A.142 build. The latest is one update later with number A.149.

What's probably also worth mentioning is that e.g. if a user flashes the second to latest one from the factory package (which flashes slot A), then does an OTA update to the latest one, they will have the latest one only in slot B. When we flash Ubuntu Touch over it we only use the _a version, which will then still be the previous version.

Long-term I'm aiming for flashing also a fixed vendor image in the installer but some open points still need to be clarified there. Until then I think the current situation is good enough (and I believe very similar to the other devices with A/B slots)

**Comment by Flohack74** [comment]
I think this is a Go! We can finetune it later, but lets get stuff done and integrate it.

**Review by Flohack74 — APPROVED** [review]

**Pull request merged**